smac_planner/src/a_star.cpp

```diff
@@ -484,6 +599,36 @@ unsigned int & AStarAlgorithm<NodeT>::getSizeDim3()
   return _dim3_size;
 }
 
+template<typename NodeT>
+typename AStarAlgorithm<NodeT>::NodePtr AStarAlgorithm<NodeT>::tryAnalyticExpansion(
+  const NodePtr & current_node, const NodeGetter & getter, int & analytic_iterations,
+  int & closest_distance)
+{
+  if (_motion_model == MotionModel::DUBIN || _motion_model == MotionModel::REEDS_SHEPP) {
+    // This must be a NodeSE2 node if we are using these motion models
+
+    // Then check if it is collision free.
+    // Always run the expansion on the first run in case there is a
+    // trivial path to be found
+    if (analytic_iterations <= 0) {
+      // Reset the counter, and try the analytic path expansion
+      analytic_iterations = closest_distance;
+      return getAnalyticPath(current_node, getter);
+    }
+    analytic_iterations--;
+    // See if we are closer and should be expanding more often
+    const Coordinates node_coords =
```

Comment:
So this logic is a little hard for me to review - you mention it runs 4-5 times from your testing, but I'm curious how that changes based on the length of the paths and where those 4-5 attempts are. In particular, it runs the first time, but as we're approaching the goal, I don't have a good intuition for how often this would run when we're in a "reasonable" proximity to be getting a valid path (e.g. if we're within lets say 20-30 cells). For longer paths, is it also 4-5 runs just spread out more or is that number proportional to distance? Just not easy questions for me to answer off hand because of the nature of how its computed. 

Another option, and I'm not suggesting you change to this right now, would be to set some number `N` as a parameter for how often to run, and then `N` is scaled by the ratio of the `current_heuristic / first-node heuristic`. I'm all for removing parameters when possible but I guess I don't have a good understanding if this is a general enough solution in case someone needs to run more (or would they want to?)

Also in the optimizations PR, the `getHeuristic` is now scaled by the neutral cost which will impact this computation.

Reply:
https://github.com/SteveMacenski/navigation2/blob/nav2_smac_planner/smac_planner/src/node_se2.cpp#L267 actually this was added in the penalties PR yesterday. You may need to pull in the nav2_smac branch and then divide by `NodeT::neutral_cost` to get back out distance (or split out the getMotionDistance() function from the getHueristicCost function()).

Reply:
You'll also be the first to see how fast things are with all this combined!

Reply:
I think this is going to be the last thing to sort out. I have changed it so that is based on the distance to the goal (removed the neutral scaling again).

The intuition is that if the current closest node is 100 units away from the goal, we try an analytic expansion in 100 steps. If we were on the optimal path with the A* expansion then this would be when we got to the goal anyway, so maybe this is incorrect.

I agree that we should probably have a way to control the frequency with which it tries to expand. The likelihood of a successful analytic expansion is obviously dependent on the type of environment (more likely in open areas), so allowing the user to tune this for their application is probably a good idea. What do you think?

Reply:
if the closest is 100 units and we search in 100 steps, conceptually, it could be perfectly trying to attempt right at the goal, which I think would happen if the heuristic and costs were "perfect". In a non-perfect world, we'd then be trying only right up close to the goal, which could lose out on trying 50 steps away which might be in direct visibility. For illustrative purposes, if we tried again in 100 steps, and at that point we're only 10 away because things aren't perfect, then we can get that final path, but I think we should try more periodically so that we can attempt on approach as well. 

I think though the only way to know what the right answer is is to test both and see which has better planning times, since this does have non-trivial overhead. 

> I agree that we should probably have a way to control the frequency with which it tries to expand. The likelihood of a successful analytic expansion is obviously dependent on the type of environment (more likely in open areas), so allowing the user to tune this for their application is probably a good idea. What do you think?

I think that's potentially an argument for a parameter `N`. Where we plan every N measurements, and its scaled by the heuristic so that as it draws smaller, the number of iterations between attempts drops as well. But like I mentioned above, I think the only way to really to know what helps more is by testing. You could test this pretty easily inline by just setting `N` as a constant in the A* file and keep the `first_heuristic` value and scale `N` by `curr_heuristic / first_heuristic`. Tune for `N` and see if we get "better" paths or faster to compute paths with some range of `N`.

Reply:
I have added a parameter to allow the user to control the frequency of expansion attempts.